docs/proposals/zeroconf.md

```diff
@@ -0,0 +1,142 @@
+
+One wrinkle is that Akri expects discovery to run to completion. Akri periodically reruns discovery for a protocol. The `zeroconf` crate polls networks for hosts and services.
+
+The implementation used by the protocol is to poll for 5 seconds and report back whichever hosts and services were discovered during that window.
+
+## Broker interfacing
+
+Upon detection of ZeroConf hosts and services, the Akri ZeroConf protocol creates "twins" for each service using the provided, sample broker. A more complete rendition of the `freddie` service could be:
+
+```YAML
+{
+    name: "freddie",
+    kind: "_example._tcp",
+    domain: "local",
+    host_name: "freddie.local",
+    address: "192.168.1.100", port: 8888,
+    txt: ...
+}
+```
+
+Each Broker Instance is configured with environment variables corresponding to the above value:
+
+```bash
+AKRI_ZEROCONF_DEVICE_KIND=_example._tcp
+AKRI_ZEROCONF=zeroconf
+AKRI_ZEROCONF_DEVICE_HOST=freddie.local
+AKRI_ZEROCONF_DEVICE_NAME=freddie
+AKRI_ZEROCONF_DEVICE_PORT=8888
+AKRI_ZEROCONF_DEVICE_ADDR=192.168.1.100
+```
+
+Currently the service's TXT records are not provided.
+
+The Broker sample enumerates these values to standard output every 5 seconds.
+
+In practice, a Kubernetes application would use this data to identify services and invoke them.
+
+## Security Considerations
+
+The Akri Agent is only able to browse services accessible to its cluster's hosts.
+
+There are no security considerations for service (mDNS) browsing. This functionality is akin to DNS lookups.
+
+Accessing services that are found by service browsing *may* require the provision of credentials. This would be an implementation detail of an Akri Broker and is not considered further here.
+
+## Outstanding Questions
+
++ What would a generic Akri ZeroConf Broker do? In practice, the application developer would likely wish to implement the Broker for their specific application.
+
+The limit of a generic Akri ZeroConf Broker is to enumerate services that it discovers and this is demonstrated by a sample Broker included in the ZeroConf Protocol implementation. In practice, a ZeroConf Broker would need to be aware of the implementation(s) of the service that it "twins". There is a limitation on the transport protocols that are accessible to Akri (discussed below) but, for the permitted transport protocols, there are potentially limitless service types and implementation details and these are all potentially accessible to an Akri Broker using this ZeroConf protocol implementation.
+
++ How to treat support for Kubernetes-supporting service types (TCP, UDP, SCTP)?
+
+Kubernetes supports [TCP, UDP, SCTP](https://kubernetes.io/docs/concepts/services-networking/service/#protocol-support) transport protocols. Service discovery supports other transport protocols. Because Akri is dependent on Kubernetes, the Akri ZeroConf Protocol implementation only supports these 3 transport protocols too.
```

Comment:
Cant the broker communicate with the device over whatever protocol but advertise the results/REST/gRPC service over TCP, UDP, SCTP service? I dont see the fact that Kubernetes services can only be those three affecting which protocol the broker uses to communicate with the device it is interfacing with.

Reply:
You're right. Good point!

Reply:
Fixed: https://github.com/deislabs/akri/pull/162/commits/5b0fed25e77c2971254125bca95b9b01f50f66d3